lib/graphql/stitching/composer/validate_resolvers.rb

```diff
@@ -12,81 +12,82 @@ def perform(supergraph, composer)
-        supergraph.fields_by_type_and_location[type.graphql_name].each do |location, field_names|
-          if field_names.none? { resolver_keys.include?(_1) }
-            raise Composer::ValidationError, "A resolver key is required for `#{type.graphql_name}` in #{location} to join with other locations."
+        # All locations of a merged type must include at least one resolver key
+        supergraph.fields_by_type_and_location[type.graphql_name].each_key do |location|
+          if resolver_keys.none? { _1.locations.include?(location) }
+            raise ValidationError, "A resolver key is required for `#{type.graphql_name}` in #{location} to join with other locations."
           end
         end
 
         # verify that all outbound locations can access all inbound locations
         resolver_locations = resolvers_by_location_and_key.keys
-        candidate_types_by_location.each_key do |location|
+        subgraph_types_by_location.each_key do |location|
           remote_locations = resolver_locations.reject { _1 == location }
           paths = supergraph.route_type_to_locations(type.graphql_name, location, remote_locations)
           if paths.length != remote_locations.length || paths.any? { |_loc, path| path.nil? }
-            raise Composer::ValidationError, "Cannot route `#{type.graphql_name}` resolvers in #{location} to all other locations. "\
+            raise ValidationError, "Cannot route `#{type.graphql_name}` resolvers in #{location} to all other locations. "\
               "All locations must provide a resolver query with a joining key."
           end
         end
       end
 
-      def validate_as_shared(supergraph, type, candidate_types_by_location)
+      def validate_as_shared(supergraph, type, subgraph_types_by_location)
         expected_fields = begin
           type.fields.keys.sort
         rescue StandardError => e
           # bug with inherited interfaces in older versions of GraphQL
           if type.interfaces.any? { _1.is_a?(GraphQL::Schema::LateBoundType) }
-            raise Composer::ComposerError, "Merged interface inheritance requires GraphQL >= v2.0.3"
+            raise CompositionError, "Merged interface inheritance requires GraphQL >= v2.0.3"
           else
             raise e
           end
         end
 
-        candidate_types_by_location.each do |location, candidate_type|
-          if candidate_type.fields.keys.sort != expected_fields
-            raise Composer::ValidationError, "Shared type `#{type.graphql_name}` must have consistent fields across locations, "\
+        subgraph_types_by_location.each do |location, subgraph_type|
+          if subgraph_type.fields.keys.sort != expected_fields
+            raise ValidationError, "Shared type `#{type.graphql_name}` must have consistent fields across locations, "\
               "or else define resolver queries so that its unique fields may be accessed remotely."
           end
```

Comment:
Need additional validations to match parsed key type structures with parsed arguments.